[docs][ci skip] Top and Testharness info (#256)
ucb-bar/chipyard: docs/Generators/Top-Testharness.rst

> Tops, TestHarnesses, and TestDriver

Reviewer: Can we enforce "Test Harness" "Test Driver"?

Reviewer: You should state that "Test Driver == Testbench"

====================================

The three highest levels of hierarchy in a Chipyard
SoC are the Top, TestHarness, and the Testbench. The Top and TestHarness are both emitted by Chisel generators.

Top/DUT
-------------------------

> The top-level module of a Rocketchip SoC is composed via cake-pattern. Specifically, "Tops" extend a "System", which extends a "Subsystem", which extends a "BaseSubsystem"

Reviewer: Rocket Chip

Reviewer: I think this should be ` for code no (not "System")

BaseSubsystem
^^^^^^^^^^^^^^^^^^^^^^^^^

> The BaseSubsystem is defined in ``generators/rocketchip/src/main/scala/subsystem/BaseSubsystem.scala``. Looking at the BaseSubsystem abstract class, we see that this class instantiates the top-level buses (frontbus, systembus, peripherybus, etc.), but does not specify a topology. We also see this class define several "ElaborationArtefacts", for example the device tree string, and the diplomacy GraphML.

Reviewer: Can you make this a link to the code also?

Reviewer: I would also make BaseSubsystem be a code word.

Author: What do you mean "link to the code". Like a hyperlink?

Author: I think RocketChip structure will change soon so this link will be dead.

Tops, Test-Harnesses, and the Test-Driver
====================================

The three highest levels of hierarchy in a Chipyard
SoC are the Top, TestHarness, and the TestDriver.
The Top and TestHarness are both emitted by Chisel generators.
The TestDriver serves as our testbench, and is a verilog

> file n rocketchip.

Reviewer 2: n rocketchip?

Top/DUT
-------------------------
The top-level module of a Rocket Chip SoC is composed via cake-pattern. Specifically, "Tops" extend a ``System``, which extends a ``Subsystem``, which extends a ``BaseSubsystem``

BaseSubsystem
^^^^^^^^^^^^^^^^^^^^^^^^^

> The ``BaseSubsystem`` is defined in ``generators/rocketchip/src/main/scala/subsystem/BaseSubsystem.scala``. Looking at the ``BaseSubsystem`` abstract class, we see that this class instantiates the top-level buses (frontbus, systembus, peripherybus, etc.), but does not specify a topology. We also see this class define several ``ElaborationArtefacts``, for example the device tree string, and the diplomacy GraphML.

Reviewer: ```suggestion
The ``BaseSubsystem`` is defined in ``generators/rocketchip/src/main/scala/subsystem/BaseSubsystem.scala``. Looking at the ``BaseSubsystem`` abstract class, we see that this class instantiates the top-level buses (frontbus, systembus, peripherybus, etc.), but does not specify a topology. We also see this class define several ``ElaborationArtefacts``, files emitted after Chisel elaboration (e.g. the device tree string, and the diplomacy GraphML).
```

Subsystem
^^^^^^^^^^^^^^^^^^^^^^^^^

> Looking in ``generators/utilities/src/main/scala/Subsystem.scala``, we can see how Chipyard extends the ``BaseSubsystem`` abstract class. The ``HasBoomAndRocketTiles`` trait defines and instantiates BOOM or Rocket tiles, depending on the parameters specified. We also connect connect some basic IOs for each tile here, specifically the hartids and the reset_vector.

Reviewer 2: connect connect?

> Looking in ``generators/utilities/src/main/scala/Subsystem.scala``, we can see how Chipyard extends the ``BaseSubsystem`` abstract class. The ``HasBoomAndRocketTiles`` trait defines and instantiates BOOM or Rocket tiles, depending on the parameters specified. We also connect connect some basic IOs for each tile here, specifically the hartids and the reset vector.

Reviewer: ```suggestion
Looking in ``generators/utilities/src/main/scala/Subsystem.scala``, we can see how Chipyard's ``Subsystem`` extends the ``BaseSubsystem`` abstract class. ``Subsystem`` mixes in the ``HasBoomAndRocketTiles`` trait that defines and instantiates BOOM or Rocket tiles, depending on the parameters specified. We also connect some basic IOs for each tile here, specifically the hartids and the reset vector.
```

System
^^^^^^^^^^^^^^^^^^^^^^^^^

> ``generators/utilities/src/main/scala/System.scala`` completes the definition of the System.

Reviewer: ```suggestion
``generators/utilities/src/main/scala/System.scala`` completes the definition of the ``System``.
```

> - ``HasHierarchicalBusTopology`` is defined in Rocket Chip, and specifies connections between the top-level buses.

Reviewer: ```suggestion
- ``HasHierarchicalBusTopology`` is defined in Rocket Chip, and specifies connections between the top-level buses
```

- ``HasAsyncExtInterrupts`` and ``HasExtInterruptsModuleImp`` adds IOs for external interrupts and wires them appropriately to tiles
- ``CanHave...AXI4Port`` adds various Master and Slave AXI4 ports, adds TL-to-AXI4 converters, and connects them to the appropriate buses
- ``HasPeripheryBootROM`` adds a BootROM device

Tops
^^^^^^^^^^^^^^^^^^^^^^^^^

A SoC top then extends the System class with any config-specific components. There are two "classes" of Tops in Chipyard.

> - ``Top`` is the default setup. These top modules instantiate a serial module which interfaces with the TestHarness. In addition, the debug module interface is tied off. All other example "Tops" (except the TopWithDTM) extend this Top as the "base"

Reviewer 2: period at the end of the sentence?

> - ``Top`` is the default setup. These top modules instantiate a serial module which interfaces with the TestHarness. In addition, the debug module interface is tied off. All other example "Tops" (except the ``TopWithDTM``) extend this Top as the "base"

Reviewer: ```suggestion
- ``Top`` is the default setup. These top modules instantiate a serial module which interfaces with the ``TestHarness``. In addition, the debug module interface is tied off. All other example "Tops" (except the ``TopWithDTM``) extend this Top as the "base"
```

- ``TopWithDTM`` does not include the serial module, and the debug module interface (DMI) is left available. This Top is used for DTM-based bringup.

> For a custom Top, for example the ``TopWithGPIO``, a mixed-in trait adds the additional modules or IOs.

Reviewer: ```suggestion
For a custom Top a mixed-in trait adds the additional modules or IOs (an example of this is ``TopWithGPIO``).
```

TestHarness
-------------------------

> There are two variants of TestHarness generators in Chipyard, both located in ``generators/example/src/main/scala/TestHarness.scala``. One is designed for TSI-based bringup, while the other performs DTM-based bringup. See TODO for more information on these two methodologies.

Reviewer 2: TODO?

> There are two variants of ``TestHarness`` generators in Chipyard, both located in ``generators/example/src/main/scala/TestHarness.scala``. One is designed for TSI-based bringup, while the other performs DTM-based bringup. See TODO for more information on these two methodologies.

Reviewer: Can you make this a link to the code.

> The wiring between the TestHarness and the Top are performed in methods defined in mixins added to the Top. When these methods are called from the TestHarness, they may instantiate modules within the scope of the harness, and then connect them to the dut. For example, the connectSimAXIMem method defined in the ``CanHaceMasterAXI4MemPortModuleImp`` trait, when called from the TestHarness, will instantiate SimAXIMems and connect them to the correct IOs of the top.

Reviewer 2: wirings? (plural)

Reviewer 2: DUT or ``dot``?

Reviewer 2: ``connectSimAxiMem``

> The wiring between the ``TestHarness`` and the Top are performed in methods defined in mixins added to the Top. When these methods are called from the TestHarness, they may instantiate modules within the scope of the harness, and then connect them to the dut. For example, the ``connectSimAXIMem`` method defined in the ``CanHaceMasterAXI4MemPortModuleImp`` trait, when called from the TestHarness, will instantiate SimAXIMems and connect them to the correct IOs of the top.

Reviewer: ```suggestion
The wiring between the ``TestHarness`` and the Top are performed in methods defined in mixins added to the Top. When these methods are called from the ``TestHarness``, they may instantiate modules within the scope of the harness, and then connect them to the DUT. For example, the ``connectSimAXIMem`` method defined in the ``CanHaveMasterAXI4MemPortModuleImp`` trait, when called from the ``TestHarness``, will instantiate SimAXIMems and connect them to the correct IOs of the top.
```

> While this roundabout way of attaching to the IOs of the top may see unnecessarily complex, it allows the designer to compose custom traits together without having to worry about the details of the implementation of any particular trait.

Reviewer 2: seem

Reviewer: I dont see this as resolved.

Specifying a Top
^^^^^^^^^^^^^^^^^^^^^^^^^

> To see why the Top connection method is useful, consider the case where we want to use a custom Top with additional GPIO pins. In ``generators/example/src/main/scala/Top.scala``, we can see how the ``TopWithGPIO`` class adds the ``HasPeripheryGPIO`` trait. This trait adds IOs to the top module, instantiates a TileLink GPIO node, and connects it to the proper buses.

Reviewer: similarly, can we point to the code.

Reviewer: Also "to the Top module"
